### PR TITLE
Oxidize openssl

### PR DESCRIPTION
**Pull request by YannBonnemay**
## Changes in this pull request

In this PR is try to remove some of the the Openssl dependance and replace it with Ring and Rustls. The motivation was some conversations with E. Scouten and A. Parsons (and also simple curiosity).

The uprooting of Openssl is not complete : I focused on signing and verifying, but left out the specific topic of ocsp checks (except for the simple "chain order" ones).

Of note :

- Rustls/Ring do not support signing/validating in all cases. As an example, you cannot verify RSA PSS with it: <https://github.com/briansmith/ring/issues/1353>, even if you can sign with it You can sign with it (<https://github.com/briansmith/ring/issues/205>). It can be a little bit confusing (<https://github.com/briansmith/ring/issues/850>).

- I did not implement OCSP checks. it seemed there was Some Assembly Required (https://github.com/rustls/rustls/issues/31, https://github.com/briansmith/webpki/issues/26).

Nice side effect : no more algorithm specific (rsa/ec/ed) structs : only one rustls signer and one validator. I think this could in turn lead to some interesting refacto in the rest of the code, but left it at that for now. 
I put the changes behind a with_rustls feature flag (default is on).

All in all, this wants to show how oxidizing Openssl in c2pa could look like, and try to be a draft for such PRs : this is a demo PR that has not been extensively tested.

## Checklist
- [] This PR represents a single feature, fix, or change.
- [x] All applicable changes have been documented.
- [x] Any `TO DO` items (or similar) have been entered as GitHub issues and the link to that issue has been included in a comment.

### CODE REVIEW

File: sdk/tests/fixtures/rustls/es256_private_key.pem [hunk 1]
```diff
@@ -0,0 +1,5 @@
+-----BEGIN PRIVATE KEY-----
```
Comment (YannBonnemay) [inline sdk/tests/fixtures/rustls/es256_private_key.pem:1]: That mass of test files felt a little loud, so I put them in their own directory

File: sdk/src/error.rs [hunk 2]
```diff
@@ -224,6 +224,22 @@ pub enum Error {
     #[cfg(feature = "file_io")]
     OpenSslError(#[from] openssl::error::ErrorStack),
 
+    #[error(transparent)]
```
Comment (YannBonnemay) [inline sdk/src/error.rs:227]: Here I added as need arised as I was a bit unsure on what are the policies on c2pr error management.

File: sdk/src/cose_validator.rs [hunk 3]
```diff
@@ -1001,6 +1014,58 @@ pub mod tests {
         }
     }
 
+    #[test]
+    #[cfg(feature = "with_rustls")]
+    fn test_verify_cose_good_with_rustls() {
```
Comment (YannBonnemay) [inline sdk/src/cose_validator.rs:1019]: These use the new fixtures.

File: sdk/tests/scripts/generate_fixtures.py [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+        "private_key": os.path.join(fixtures_dir, "%s_private_key.pem" % test_id),
+        "public_key": os.path.join(fixtures_dir, "%s_public_key.pem" % test_id),
+        "public_key_der": os.path.join(fixtures_dir, "%s_public_key.der" % test_id),
+        "signed_data": os.path.join(fixtures_dir, "%s_signed_data.sign" % test_id),
+        "fixtures_data": os.path.join(fixtures_dir, "../data.data"),
+    }, **FIXTURES_SPECIFICS[test_id]}
+
+def generate_fixtures_rsa(data):
+    cmds = [
+        "openssl genpkey -algorithm {algorithm} -out {private_key}".format(**data), # generate private key
```
Comment (YannBonnemay) [inline sdk/tests/scripts/generate_fixtures.py:75]: The actual test fixtures generation is actually a side effect, I think : this was documentation for me.

File: sdk/src/openssl/ec_signer.rs [hunk 5]
```diff
@@ -243,6 +243,7 @@ mod tests {
     }
 
     #[test]
+    #[cfg_attr(not(feature = "with_rustls"), ignore)]
```
Comment (YannBonnemay) [inline sdk/src/openssl/ec_signer.rs:246]: These are `feature`d  out when `with_rustls`, so I had to add that here.

File: sdk/src/rustls/common.rs [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+    iter,
+};
+
+pub struct AlgorithmData {
+    pub rustls_id: rustls::SignatureScheme,
+    pub verification_alg: &'static dyn ring::signature::VerificationAlgorithm,
+    // This is because the Ring verification algorithm expects the key in a SubjectPublicKey, Openssl gives it in a SubjectPublicKeyInfo.
+    // We can ask Openssl to give us that with RSAPublicKey_out, but that's (i) only for RSA, (ii) not what happens in existing tests.
+    // https://github.com/briansmith/ring/issues/881#issuecomment-592749266
+    pub spk_offset: usize,
```
Comment (YannBonnemay) [inline sdk/src/rustls/common.rs:28]: Flagging that hack here. As the structures are well defined, it should be safe to just fast forward to the bytes where the usable part is.

Still.

File: sdk/src/rustls/temp_signer.rs [hunk 7]
```diff
@@ -0,0 +1,468 @@
...
+// This function panics if unable to invoke openssl as expected.
+//
+fn create_x509_key_pair(
+    sign_cert_path: &Path,
+    pem_key_path: &Path,
+    has_priv_key: bool,
+    rsa_padding_mode: Option<&str>,
+    sha_mode: Option<&str>,
+) {
+    let mut openssl = Command::new("openssl");
```
Comment (YannBonnemay) [inline sdk/src/rustls/temp_signer.rs:106]: I took care of the openssl uses in the Rust code, but and did not change these calls to it.

File: sdk/src/rustls/validator.rs [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+    }
+}
+
+impl CoseValidator for Validator {
+    fn validate(&self, sig: &[u8], data: &[u8], pkey: &[u8]) -> Result<bool> {
+        let algorithm_data = &get_algorithm_data(&self.alg).map_err(|_err| Error::CoseSignature)?;
+
+        let public_key = signature::UnparsedPublicKey::new(
+            algorithm_data.verification_alg,
+            &pkey[algorithm_data.spk_offset..],
```
Comment (YannBonnemay) [inline sdk/src/rustls/validator.rs:36]: Offset used here (see comment in common.rs)

File: sdk/src/cose_validator.rs [hunk 9]
```diff
@@ -1017,25 +1082,25 @@ pub mod tests {
     #[test]
     fn test_verify_ec_good() {
         // EC signatures
-        let mut validator = get_validator("es384").unwrap();
+        // let validator = get_validator("es384").unwrap();
 
-        let sig_es384_bytes = include_bytes!("../tests/fixtures/sig_es384.data");
-        let data_es384_bytes = include_bytes!("../tests/fixtures/data_es384.data");
-        let key_es384_bytes = include_bytes!("../tests/fixtures/key_es384.data");
+        // let sig_es384_bytes = include_bytes!("../tests/fixtures/sig_es384.data");
```
Comment (YannBonnemay) [inline sdk/src/cose_validator.rs:1087]: This one is a little mystery - I can't have these tests pass using the Ring backent, and I can't seem to be able to validate them with command line openssl (`openssl dgst -sha256 -verify ../tests/fixtures/key_es384.data -signature ../tests/fixtures/sig_es384.data ../tests/fixtures/data_es384.data`).

My generated tests, otoh, work ok with both Openssl and Ring validation.

Left commented version to flag the oddity.

Comment (mauricefisher64) [reply sdk/src/cose_validator.rs:1087]: Do you handle IEEE P1363 signature format?  That is what is spec'd by C2PA

File: sdk/src/validator.rs [hunk 10]
```diff
@@ -59,24 +60,28 @@ impl CoseValidator for DummyValidator {
 // • PS512 (RSASSA-PSS using SHA-512 and MGF1 with SHA-512)
 // • RS256	RSASSA-PKCS1-v1_5 using SHA-256
 // • RS384	RSASSA-PKCS1-v1_5 using SHA-384
-// • RS512	RSASSA-PKCS1-v1_5 using SHA-512
+// • RS512	RSASSA-PKCS1-v1_5 using SHA-512 // validation not supported by Ring
 // • ED25519 Edwards Curve ED25519
```
Comment (mauricefisher64) [inline sdk/src/validator.rs:64]: All RSAxxx have been deprecated by C2PA so it OK to remove these.

File: sdk/tests/scripts/generate_fixtures.py (hunk 4 above)
Comment (mauricefisher64) [reply sdk/tests/scripts/generate_fixtures.py:75]: We now provide sample certs for all signature types as to remove the need for a developer just trying the SDK from needing OpenSLL to generate certs.

File: sdk/src/cose_validator.rs (hunk 9 above)
Comment (YannBonnemay) [reply sdk/src/cose_validator.rs:1087]: Complete oversight on my part. Should be possible to re-add it, inasmuch as the packing/unpacking seems to happens outside of openssl (seems to be the case?).

File: sdk/src/validator.rs (hunk 10 above)
Comment (YannBonnemay) [reply sdk/src/validator.rs:64]: Excellent, I'll remove that.

File: sdk/tests/scripts/generate_fixtures.py (hunk 4 above)
Comment (YannBonnemay) [reply sdk/tests/scripts/generate_fixtures.py:75]: Added to the chopping block.

File: sdk/src/rustls/rustls_signer.rs [hunk 11]
```diff
@@ -0,0 +1,208 @@
...
+        signcert: &[u8],
+        pkey: &[u8],
+        alg: String,
+        tsa_url: Option<String>,
+    ) -> Result<Self> {
+        // Get signcerts, the certificates vector
+        let signcerts = get_certificates(signcert);
+        let pkey = get_private_key(&alg, pkey)?;
+
+        check_chain_order(&signcerts);
```
Comment (mauricefisher64) [inline sdk/src/rustls/rustls_signer.rs:83]: Any reason to call this twice

File: sdk/tests/fixtures/rustls/es256_private_key.pem (hunk 1 above)
Comment (YannBonnemay) [reply sdk/tests/fixtures/rustls/es256_private_key.pem]: Removed

File: sdk/tests/scripts/generate_fixtures.py (hunk 4 above)
Comment (YannBonnemay) [reply sdk/tests/scripts/generate_fixtures.py]: New tests were a god send. Thanks!

File: sdk/src/rustls/common.rs (hunk 6 above)
Comment (YannBonnemay) [reply sdk/src/rustls/common.rs]: Removed

File: sdk/src/rustls/validator.rs (hunk 8 above)
Comment (YannBonnemay) [reply sdk/src/rustls/validator.rs]: Removed

File: sdk/src/cose_validator.rs (hunk 9 above)
Comment (YannBonnemay) [reply sdk/src/cose_validator.rs]: IEEE P1363 managemen added

File: sdk/src/rustls/rustls_signer.rs (hunk 11 above)
Comment (YannBonnemay) [reply sdk/src/rustls/rustls_signer.rs]: None, removed

**Comment by mauricefisher64** [comment]
OCSP is about to be enabled so if you can find a solution for that it would be great.  Also trust list support is also coming so if you have a way to validate a certificate chain against a set of trust anchors that would be good.  Does this PR work with WASM?

**Pull request closed without merging**